### ISSUE TITLE
NPE at aQute.bnd.osgi.Macro.process(Macro.java:96)

### ISSUE THREAD

**Issue by xtracoder** [bndtools/bnd#1270]
I'm getting NullPointerException by stacktrace below 
(Bnd is built from sources at revision 
SHA-1: 5780ccf68be8d0365f84ff385ed93632e6e17f5a
Merge pull request #1268 from bjhargrave/javac.defaults)

```
"main"
	at java.lang.StringBuilder.<init>(StringBuilder.java:125)
	at aQute.bnd.osgi.Macro.process(Macro.java:96)
	at aQute.bnd.osgi.Macro.process(Macro.java:91)
	at aQute.bnd.osgi.Macro.process(Macro.java:86)
	at aQute.bnd.osgi.Macro.process(Macro.java:1204)
	at aQute.bnd.osgi.Macro.getFlattenedProperties(Macro.java:1152)
	at aQute.bnd.osgi.Macro.getFlattenedProperties(Macro.java:1128)
	at aQute.bnd.osgi.Processor.getFlattenedProperties(Processor.java:1375)
	at aQute.bnd.osgi.Analyzer.calcManifest(Analyzer.java:881)
	at aQute.bnd.osgi.Builder.build(Builder.java:105)
	at aQute.bnd.osgi.Builder$build.call
```

It occurs because `process()` is invoked with `null` argument
```
	public Properties getFlattenedProperties(boolean ignoreInstructions) {
	      ...
						flattened.put(key, process(source.getProperty(key)));
```

what happens because of this code
```
package java.util;

public class Properties extends Hashtable<Object,Object> {
    public String getProperty(String key) {
        Object oval = super.get(key);
        String sval = (oval instanceof String) ? (String)oval : null;
        return ((sval == null) && (defaults != null)) ? defaults.getProperty(key) : sval;
    }
```

because value stored in Hashtable is actually a GString (a Groovy String) and is evaluated as null therefore (i would say Hashtable implementation is the origin of that problem). At the moment i have no idea how GString comes into Bnd's code (project is build by Gradle + plugin: 'org.dm.bundle''), but that stack trace with NPE is absolutely misleading for problem resolution and such weird condition should be handled more gracefully within Bnd code - for example name of the 'bad' key should be logged and more informative exception can be thrown.

**Comment by bjhargrave** [comment]
I think this must be a bug in that gradle plugin. It must be putting a GString value in the properties which must be `<String,String>`. I suggest you file a bug with them.

bnd expects all values to be String. So if someone using the bnd api puts a non-String in, bnd can fail.

**Comment by xtracoder** [comment]
Yep, of course something has to be fixed in 'that' plugin and will take about it separately. I'm just saying that Bnd better be somewhat more bullet-proof in that case of 'wrong' usage.

**Comment by bjhargrave** [comment]
This is a unique problem to groovy where a GString object can be "converted" to null. It is not possible to put a null value in a Properties object. So bnd has every right to expect that it will never see a null value out of a Properties object. Adding checks to null values out of a Properties object is a bit overkill since the Properties contract says this is not possible.

**Comment by xtracoder** [comment]
The problem is actually not about GString only (groovy is dynamic language and does not perform sufficient type checks and it seems the problem is bit more severe than I expected originally - root exception occurs earlier in the code, here

```
package aQute.bnd.osgi;
public class Processor extends Domain implements Reporter, Registry, Constants, Closeable {
	public String getProperty(String key, String deflt, String separator) {
		...
				value = (String) source.getProperties().get(key);
```

what is invoked from
```
public class Analyzer extends Processor {
	public Manifest calcManifest() throws Exception {
		try {
		...
		}
		catch (Exception e) {
			// This should not really happen. The code should never throw
			// exceptions in normal situations. So if it happens we need more
			// information. So to help diagnostics. We do a full property dump
			throw new IllegalStateException("Calc manifest failed, state=\n" + getFlattenedProperties(), e);
		}
	}
```
And NPE exception occurs in the code which reports source exception :) ... and, BTW, after fixing source of GString i've got another one, with not GString:

```
Caused by: java.lang.IllegalArgumentException: Property Bnd-LastModified is not a not-null String, value: {java.lang.Long} 1450368311594
        at aQute.bnd.osgi.Macro.getFlattenedProperties(Macro.java:1155)
```

And here is the updated code (for copy/paste?), with more intelligent processing of that case 
```
public Properties getFlattenedProperties(boolean ignoreInstructions) {
        // Some macros only work in a lower processor, so we
        // do not report unknown macros while flattening
        flattening = true;
        try {
            Properties flattened = new UTF8Properties();
            Properties source = domain.getProperties();

            for( Enumeration< ?> e = source.propertyNames(); e.hasMoreElements(); ) {
                String key = (String)e.nextElement();
                if( !key.startsWith("_") ) {
                    String propValueString = source.getProperty(key);

                    if( propValueString == null ) {
                        Object propValue = source.get(key);
                        throw new IllegalArgumentException(String.format(
                                "Property %s is not a not-null String, value: %s%s",
                                key, propValue != null ? "{" + propValue.getClass().getName() + "} " : "", propValue));
                    }

                    if( ignoreInstructions && key.startsWith("-") )
                        flattened.put(key, propValueString);
                    else
                        flattened.put(key, process(propValueString));
                }
            }
            return flattened;
        }
        finally {
            flattening = false;
        }
    }
```

**Comment by xtracoder** [comment]
Ok, so the final diagnose for root cause of second use-case is - while trying to identify which 'osgi' gradle plugin better suites my needs i've occasionally got both of them in one build.script. Native 'osgi' plugin adds Long value into attributes and another plugin forwards entire collection to Bnd Tools.

Of course NPE does not give any hint to me, as an end user, about the reason :(

BTW, the piece of code i've provided for getFlattenedProperties() will not solve the final problem - reporting the exception in Analyzer.calcManifest() and i would say calling getFlattenedProperties() in the Exception handler is not a good choice - i think it's better to add another method to convert Properties into list of <key, {type if not String}value> and enumerate it via underlying Dictionary interface, which provides raw access to values.

**Comment by bjhargrave** [comment]
Even code like you propose has some issues The `Properties.getProperty` method will ignore a property with a non-String value and look in the defaults. So if there is a default with a String value, it will be "in effect" and the override with the non-String value is ignored. There is no public API to interrogate the defaults directly.

I made a fix, above, for 3.2. It does not throw exceptions. It makes a warning and acts as if the key was not present.
